PR dev/ilyasa/module_2 to module_2 14/07/2025 08:55 WIB

zettacamp-nabil-ilyasa opened the pull request:
List of Changes:
- CRUD Resolvers for Block
- CRUD Resolvers for Subject
- CRUD Resolvers for Test
- Implement Test Lificycle Flow

1. Publish Test:
<img width="1365" height="187" alt="Cuplikan layar 2025-07-14 094524" src="https://github.com/user-attachments/assets/22a68647-c24d-4cbf-ba75-47a1e4484cb8" />


2. Assign Corrector Task
<img width="1365" height="166" alt="Cuplikan layar 2025-07-14 094826" src="https://github.com/user-attachments/assets/35a93cce-1147-4117-8c35-69153e98d788" />


3. Generate Enter Marks Task for all students
<img width="1365" height="680" alt="Cuplikan layar 2025-07-14 094916" src="https://github.com/user-attachments/assets/e0b722c7-b55b-435e-865c-6f77f95768e1" />


4. SendGrid Email Notification:
<img width="1071" height="486" alt="Cuplikan layar 2025-07-14 095402" src="https://github.com/user-attachments/assets/72702db4-d063-4bee-a732-8ab41de28a3a" />


5. Enter Marks
<img width="1365" height="216" alt="Cuplikan layar 2025-07-14 095117" src="https://github.com/user-attachments/assets/d4ea2d06-0e45-44c4-9d7a-8fd0e9852b09" />


6. Enter Marks task status changed into completed, validate marks task generated
<img width="1365" height="575" alt="image" src="https://github.com/user-attachments/assets/0344d153-f015-4011-8afd-45f50a8dcc31" />


7. Validate Marks:
<img width="1365" height="143" alt="image" src="https://github.com/user-attachments/assets/019226b6-b780-4349-aa0d-38964ace2fb8" />


8.  Validate Marks task completed:
<img width="1271" height="454" alt="image" src="https://github.com/user-attachments/assets/1030aac7-08f3-4aec-b17c-dd9f8c8f9214" />


9. Student test result validated:
<img width="1365" height="358" alt="image" src="https://github.com/user-attachments/assets/0c76409a-0ef9-449b-a8d9-4b388d529818" />

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_001] Implement CRUD Resolvers for Block Model : Create Block Model

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_002] Implement CRUD Resolvers for Block Model : Create Block Typedef

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_003] Implement CRUD Resolvers for Block Model : Create Block Validator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_004] Implement CRUD Resolvers for Block Model : Create Query GetAllBlocks

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_005] Implement CRUD Resolvers for Block Model : Create Query GetOneBlock

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_006] Implement CRUD Resolvers for Block Model : Create Mutation CreateBlock

- add BlockPayloadComposer helper

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_007] Implement CRUD Resolvers for Block Model : Create Mutation UpdateBlock

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_008] Implement CRUD Resolvers for Block Model : Create Mutation DeleteBlock

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_009] Implement CRUD Resolvers for Block Model : Create BlockLoader

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_010] Implement CRUD Resolvers for Block Model : Create Subject Model

- +minor revision block model

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_011] Implement CRUD Resolvers for Subject Model : Create Subject Typedef

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_012] Implement CRUD Resolvers for Subject Model : Create Subject Validator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_011] Implement CRUD Resolvers for Subject Model: Create Subject Typedef

- revise coefficient type

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_003] Implement CRUD Resolvers for Block Model : Create Block Validator

- minor edit for jsdoc

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_013] Implement CRUD Resolvers for Subject Model : Create Query GetAllSubjects

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_005] Implement CRUD Resolvers for Block Model : Create Query GetOneBlock

- minor fix on the return

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_014] Implement CRUD Resolvers for Subject Model : Create GetOneSubject

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_015] Implement CRUD Resolvers for Subject Model : Create Mutation CreateSubject

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_016] Implement CRUD Resolvers for Subject Model : Create Mutation UpdateSubject

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_017] Implement CRUD Resolvers for Subject Model : Create Mutation DeleteSubject

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_006] Implement CRUD Resolvers for Block Model : Create Mutation CreateBlock

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_018] Implement CRUD Resolvers for Subject Model : Create SubjectLoader

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_019] Implement CRUD Resolvers for Subject Model : Create PaginationValidator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_015] Implement CRUD Resolvers for Subject Model :Create Mutation CreateSubject

+existence checking for block

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_012] Implement CRUD Resolvers for Subject Model :Create Subject Validator

+ control flag for validator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_020] Implement CRUD Resolvers for Test Model : Create Test Model

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16[BE_021] Implement CRUD Resolvers for Test Model : Create Test Typedef

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_010] Implement CRUD Resolvers for Block Model : Create Subject Model

minor fix on deleted_at field

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_011] Implement CRUD Resolvers for Subject Model : Create Subject Typedef

delete status from subject typedef

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_022] Implement CRUD Resolvers for Test Model : Create Test Validator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_023] Implement CRUD Resolvers for Test Model : Create Query GetAllTests

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_024] Implement CRUD Resolvers for Test Model : Create Query GetOneTest

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_025] Implement CRUD Resolvers for Test Model : Create Mutation CreateTest

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_010] Implement CRUD Resolvers for Test Model : Create BlockPayloadComposer

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_021] Implement CRUD Resolvers for Test Model :  Create SubjectPayloadComposer

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_032] Implement CRUD Resolvers for Test Model : Create TestPayloadComposer

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_032] Implement CRUD Resolvers for Test Model : Create GetTotalWeightOfTests

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_034] Implement CRUD Resolvers for Test Model : Create TestFilterInput Validator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_034] Implement CRUD Resolvers for Test Model : Create TestFilterInput Validator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_028] Implement CRUD Resolvers for Test Model : Create Mutation UpdateTest

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_030] Implement CRUD Resolvers for Test Model : Create Mutation DeleteTest

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_031] Implement CRUD Resolvers for Test Model : Create TestLoader

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-10][BE_035] Define StudentTestResult Model: Create StudentTestResult Model

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-10][BE_035] Define StudentTestResult Model: Create StudentTestResult Model

zettacamp-nabil-ilyasa committed:
Create studentTestResult.typedef.js

[MODULE_2][ZEBN-10][BE_036] Define StudentTestResult Model: Create Type StudentTestResult

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16[BE_021] Implement CRUD Resolvers for Test Model : Create Test Typedef

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-10][BE_044] Define StudentTestResult Model: Create ValidateStudentTestResultFilterInput

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-10][BE_038] Define StudentTestResult Model: Create GetAllStudentTestResults

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-39][BE_038] Define StudentTestResult Model: Create GetOneStudentTestResult

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-10][BE_041] Define StudentTestResult Model: Create DeleteStudentTestResult

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-12][BE_045] Define Task Management Schema: Create Task Model

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-12][BE_046] Define Task Management Schema: Create Task Typedef

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-12][BE_059] Define Task Management Schema: Create ValidateTaskFilterInput

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-12][BE_048] Define Task Management Schema: Create GetAllTasks

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-12][BE_049] Define Task Management Schema: Create GetOneTask

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-12][BE_052] Define Task Management Schema: Create DeleteTask

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-12][BE_052] Define Task Management Schema: Create DeleteTask

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-12][BE_053] Define Task Management Schema: Create TaskLoader

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-20][BE_042] Implement publishTest Mutation: Create CreateAssignCorrectorTask

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-10][BE_037] Define StudentTestResult Model: Create ValidateMarks

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_009] Implement CRUD Resolvers for Block Model : BlockLoader

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_004] Implement CRUD Resolvers for Block Model : Create Query GetAllBlocks

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_005] Implement CRUD Resolvers for Block Model : Create Query GetOneBlock

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_001] Implement CRUD Resolvers for Block Model : Create Block Model

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_002] Implement CRUD Resolvers for Block Model : Create Block Typedef

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-14][BE_003] Implement CRUD Resolvers for Block Model : Create Block Validator

zettacamp-nabil-ilyasa committed:
[Module_1][CRUD School][ZEBN-5] Revise commentaries

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][BE_045] Define StudentTestResult Model: Create SetEnterMarksTaskToValidated

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][BE_046] Define StudentTestResult Model: Create GetTestNotation

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][BE_043] Define StudentTestResult Model: Create EnterMarksPayloadComposer

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][BE_037] Define StudentTestResult Model: Create EnterMarksPayloadComposer

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-10][BE_035] Define StudentTestResult Model: StudentTestResult Model

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-10][BE_035] Define StudentTestResult Model: StudentTesRtesult Model

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-19[BE_035] Implement GraphQL Queries for Task Data: GetAllStudentTestResults

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-19[BE_039] Implement GraphQL Queries for Task Data: GetOneStudentTestResult

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-19[BE_041] Implement GraphQL Queries for Task Data: DeleteStudentTestResult

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][BE_043] Implement Enter Marks: Create EnterMarks

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][BE_047] Implement Enter Marks: Create EnterMarks: Implement loader for field resolver

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-21][BE_060] Implement Validate Marks: Create Validator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-20][BE_023] Implement CRUD Resolvers for Test Model : Cr… …eate TestFilterInput Validator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-13][BE-052] Implement assignCorrector Mutation

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_013] Implement CRUD Resolvers for Subject Model: Create Query GetAllSubjects

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16[BE_021] Implement CRUD Resolvers for Test Model : Create Test Typedef

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_012] Implement CRUD Resolvers for Subject Model :Create Subject Validator

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-10][BE_036] Define StudentTestResult Model: Create ValidateStudentTestResult Typedef

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-20][BE_023] Implement CRUD Resolvers for Test Model : Publish Test

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_32] Implement CRUD Resolvers for Test Model : TestPayloadComposer

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-13][BE-052] Implement assignCorrector Mutation

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-15][BE_015] Implement CRUD Resolvers for Subject Model :Mutation CreateSubject

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-16][BE_025] Implement CRUD Resolvers for Test Model : CreateTest

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-21][BE_060] Implement Validate Marks

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][BE_043] Implement Enter Marks: Fix average marks

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][ Syllabus & Test Management] Fix PR Review

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][Syllabus & Test Management] Fix PR Review

SantikaPutra commented on modules/studenTestResult/student_test_result.model.js:
```diff
@@ -0,0 +1,61 @@
+
+  // mark or grade
+  mark: { type: Number, required: true },
+});
```
declare the object directly on the field

zettacamp-nabil-ilyasa replied:
fixed on commit 7f6c35e48c27c83f61cfcc2df14830934d40055e

SantikaPutra commented on modules/test/test.model.js:
```diff
@@ -0,0 +1,58 @@
+
+  // maximum point that can be achieved
+  max_points: { type: Number, required: true },
+});
```
declare the object directly on the field

zettacamp-nabil-ilyasa replied:
fixed on commit 7f6c35e48c27c83f61cfcc2df14830934d40055e

SantikaPutra commented on modules/block/block.helper.js:
```diff
@@ -0,0 +1,20 @@
+    throw new ApolloError('name is required for payload');
+  }
+  // *************** return composed payload
+  return { name: blockName, description: blockDescription };
```
why?

zettacamp-nabil-ilyasa replied:
wrapped the result with variable before return on commit 2e5ce88c46111c36a91ae351c95d8f0ae285b056

SantikaPutra commented on modules/subject/subject.helper.js:
```diff
@@ -0,0 +1,36 @@
+    description: inputObject.description,
+    block_id: inputObject.block_id,
+    coefficient: inputObject.coefficient,
+  };
```
why?

zettacamp-nabil-ilyasa replied:
fixed by commit 2e5ce88c46111c36a91ae351c95d8f0ae285b056

SantikaPutra commented on modules/test/test.helper.js:
```diff
@@ -0,0 +1,106 @@
+  };
+
+  // *************** validate subject_id if checkSubjectId set to true
+  if (addSubjectId) {
+    ValidateMongoObjectId(inputObject.subject_id);
+    testPayload.subject_id = inputObject.subject_id;
+  }
```
just validate on the resolver to not allow subject id to be changed

zettacamp-nabil-ilyasa replied:
force the subject id for update and create
not allowing subject_id change for update
on commit 2e5ce88c46111c36a91ae351c95d8f0ae285b056

SantikaPutra commented on modules/subject/subject.resolver.js:
```diff
@@ -0,0 +1,292 @@
+ * @returns {Promise<Object>} - Updated subject document.
+ * @throws {ApolloError} - Throws error if validation or db operation fails.
+ */
+async function UpdateSubject(parent, { _id, input }) {
```
add validation when block id is changed

zettacamp-nabil-ilyasa replied:
force the block_id for update and create
not allowing block_id change for update
on commit 2e5ce88c46111c36a91ae351c95d8f0ae285b056

SantikaPutra commented on modules/test/test.resolver.js:
```diff
@@ -0,0 +1,319 @@
+    }
+
+    // *************** check if combined tests weight is exceed 1
+    const currentWeight = await GetTotalWeightOfTests(toBeUpdatedTestDocument.subject_id);
```
use source from input to make it same like create test

zettacamp-nabil-ilyasa replied:
fixed on commit 2e5ce88c46111c36a91ae351c95d8f0ae285b056

SantikaPutra commented on modules/test/test.resolver.js:
```diff
@@ -0,0 +1,319 @@
+  }
+
+  // *************** create assign corrector task
+  const taskOwnerUserId = '6862150331861f37e4e3d209';
```
no magic string

zettacamp-nabil-ilyasa replied:
added into shared/strings.js per commit 2e5ce88c46111c36a91ae351c95d8f0ae285b056

SantikaPutra commented on modules/studenTestResult/student_test_result.helper.js:
```diff
@@ -0,0 +1,148 @@
+    average_mark: averageMark,
+    status: 'completed',
+    mark_entry_date: new Date(),
+  };
```
why?

zettacamp-nabil-ilyasa replied:
wrapped per commit 2e5ce88c46111c36a91ae351c95d8f0ae285b056

SantikaPutra commented on modules/studenTestResult/student_test_result.helper.js:
```diff
@@ -0,0 +1,148 @@
+  }
+
+  // *************** calculate average mark, ensure that average mark only has 2 decimal
+  const averageMark = Math.round((studentMarks.reduce((acc, studentMark) => acc + studentMark.mark, 0) / studentMarks.length) * 100) / 100;
```
use `Number.toFixed(2)`

zettacamp-nabil-ilyasa replied:
updated on commit 2e5ce88c46111c36a91ae351c95d8f0ae285b056

SantikaPutra commented on modules/studenTestResult/student_test_result.resolver.js:
```diff
@@ -0,0 +1,326 @@
+ * @returns {Promise<Object>} The newly created StudentTestResult document.
+ * @throws {ApolloError} If validation fails, student test result already exists, or any DB operation fails.
+ */
+async function EnterMarks(parent, { taskId, studentMarks }) {
```
no mutation to update mark?

zettacamp-nabil-ilyasa replied:
UpdateEnteredMarks added
update student test result which status = completed, by providing same input which is task_id and marks
task provided should be a 'validate_marks' task that still in-progress and stores the same id with that of StudentTestResult to ensure the Test Lifecycle flow is not broken

changes on commit: 9b6209477dc178376e622f28a0032fd75ed2a6b9

SantikaPutra commented on modules/task/task.resolver.js:
```diff
@@ -0,0 +1,418 @@
+    // *************** add due_date if provided
+    if (dueDate) {
+      updateAssignCorrectorTask.due_date = new Date(dueDate);
+    }
```
add composer for consistency

zettacamp-nabil-ilyasa replied:
added on commit 2e5ce88c46111c36a91ae351c95d8f0ae285b056

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][Syllabus & Test Management] Fix PR Review

- add shared folder containing static strings
- put mark and notation object straight into the model without creating new schema

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][Syllabus & Test Management] Fix PR Review

- add UpdateEnteredMarks
- add payload composer helper for task

zettacamp-nabil-ilyasa committed:
[MODULE_2][ZEBN-18][Syllabus & Test Management] Fix PR Review

SantikaPutra reviewed (approved):
Approved
